File: python/semantic_kernel/diagnostics/validation_exception.py
# Copyright (c) Microsoft. All rights reserved.

from enum import Enum
from typing import Optional

from semantic_kernel.diagnostics.sk_exception import SKException


class ValidationException(SKException):
    class ErrorCodes(Enum):
        # Unknown error.
        UnknownError = -1
        # Null value.
        NullValue = 0
        # Empty value.
        EmptyValue = 1
        # Out of range.
        OutOfRange = 2
        # Missing prefix.
        MissingPrefix = 3
        # Directory not found.
        DirectoryNotFound = 4

    # The error code.
    _error_code: ErrorCodes

    def __init__(
        self,
        error_code: ErrorCodes,
        message: str,
        inner_exception: Optional[Exception] = None,
    ) -> None:
        """Initializes a new instance of the ValidationException class.

        Arguments:
            error_code {ErrorCodes} -- The error code.
            message {str} -- The error message.
            inner_exception {Exception} -- The inner exception.
        """
        super().__init__(error_code, message, inner_exception)
        self._error_code = error_code

    @property
    def error_code(self) -> ErrorCodes:
        """Gets the error code.

        Returns:
            ErrorCodes -- The error code.
        """
        return self._error_code

File: dotnet/src/SemanticKernel.Abstractions/Diagnostics/ValidationException.cs
// Copyright (c) Microsoft. All rights reserved.

using System;

namespace Microsoft.SemanticKernel.Diagnostics;

#pragma warning disable CA1032 // Implement standard exception constructors

/// <summary>
/// Exception thrown for errors related to validation.
/// </summary>
public class ValidationException : SKException
{
    /// <summary>
    /// Initializes a new instance of the <see cref="ValidationException"/> class with a provided error code.
    /// </summary>
    /// <param name="errorCode">The error code.</param>
    public ValidationException(ErrorCodes errorCode)
        : this(errorCode, message: null, innerException: null)
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ValidationException"/> class with a provided error code and message.
    /// </summary>
    /// <param name="errorCode">The error code.</param>
    /// <param name="message">The exception message.</param>
    public ValidationException(ErrorCodes errorCode, string? message)
        : this(errorCode, message, innerException: null)
    {
    }

    /// <summary>
    /// Initializes a new instance of the <see cref="ValidationException"/> class with a provided error code, message, and inner exception.
    /// </summary>
    /// <param name="errorCode">The error code.</param>
    /// <param name="message">A string that describes the error.</param>
    /// <param name="innerException">The exception that is the cause of the current exception.</param>
    public ValidationException(ErrorCodes errorCode, string? message, Exception? innerException)
        : base(GetDefaultMessage(errorCode, message), innerException)
    {
        this.ErrorCode = errorCode;
    }

    /// <summary>
    /// Gets the error code for this exception.
    /// </summary>
    public ErrorCodes ErrorCode { get; }

    /// <summary>Translate the error code into a default message.</summary>
    private static string GetDefaultMessage(ErrorCodes errorCode, string? message)
    {
        string description = errorCode switch
        {
            ErrorCodes.NullValue => "Null value",
            ErrorCodes.EmptyValue => "Empty value",
            ErrorCodes.OutOfRange => "Out of range",
            ErrorCodes.MissingPrefix => "Missing prefix",
            ErrorCodes.DirectoryNotFound => "Directory not found",
            _ => $"Unknown error ({errorCode:G})",
        };

        return message is not null ? $"{description}: {message}" : description;
    }

    /// <summary>
    /// Error codes for <see cref="ValidationException"/>.
    /// </summary>
    public enum ErrorCodes
    {
        /// <summary>
        /// Unknown error.
        /// </summary>
        UnknownError = -1,

        /// <summary>
        /// Null value.
        /// </summary>
        NullValue,

        /// <summary>
        /// Empty value.
        /// </summary>
        EmptyValue,

        /// <summary>
        /// Out of range.
        /// </summary>
        OutOfRange,

        /// <summary>
        /// Missing prefix.
        /// </summary>
        MissingPrefix,

        /// <summary>
        /// Directory not found.
        /// </summary>
        DirectoryNotFound,
    }
}
